sw/device/lib/base/memory_unittest.cc

```diff
@@ -0,0 +1,404 @@
+  }
+}
+
+TEST_P(MemChrTest, RepeatedBytes) {
+  auto memchr_func = GetParam();
+
+  const bool reverse =
+      memchr_func == &ot_memrchr || memchr_func == &ref_memrchr;
+
+  // Define a buffer that contains repeated bytes at a variety of offsets.
+  constexpr uint8_t kBuf[] = {// No repeated bytes yet.
+                              0, 1, 2, 3,
+                              // First byte is repeated once.
+                              4, 4, 5, 6,
+                              // Second byte is repeated once
+                              7, 8, 8, 9,
+                              // Third byte is repeated once.
+                              10, 11, 12, 12,
+                              // First byte is repeated three times.
+                              13, 13, 13, 14,
+                              // Second byte is repeated three times.
+                              15, 16, 16, 16,
+                              // First byte is repeated four times.
+                              17, 17, 17, 17};
+
+  EXPECT_EQ(memchr_func(kBuf, 0, sizeof(kBuf)), kBuf);
+  EXPECT_EQ(memchr_func(kBuf, 1, sizeof(kBuf)), kBuf + 1);
+  EXPECT_EQ(memchr_func(kBuf, 2, sizeof(kBuf)), kBuf + 2);
+  EXPECT_EQ(memchr_func(kBuf, 3, sizeof(kBuf)), kBuf + 3);
+  EXPECT_EQ(memchr_func(kBuf, 4, sizeof(kBuf)), reverse ? kBuf + 5 : kBuf + 4);
+  EXPECT_EQ(memchr_func(kBuf, 5, sizeof(kBuf)), kBuf + 6);
+  EXPECT_EQ(memchr_func(kBuf, 6, sizeof(kBuf)), kBuf + 7);
+  EXPECT_EQ(memchr_func(kBuf, 7, sizeof(kBuf)), kBuf + 8);
+  EXPECT_EQ(memchr_func(kBuf, 8, sizeof(kBuf)), reverse ? kBuf + 10 : kBuf + 9);
+  EXPECT_EQ(memchr_func(kBuf, 9, sizeof(kBuf)), kBuf + 11);
+  EXPECT_EQ(memchr_func(kBuf, 10, sizeof(kBuf)), kBuf + 12);
+  EXPECT_EQ(memchr_func(kBuf, 11, sizeof(kBuf)), kBuf + 13);
+  EXPECT_EQ(memchr_func(kBuf, 12, sizeof(kBuf)),
+            reverse ? kBuf + 15 : kBuf + 14);
+  EXPECT_EQ(memchr_func(kBuf, 13, sizeof(kBuf)),
+            reverse ? kBuf + 18 : kBuf + 16);
+  EXPECT_EQ(memchr_func(kBuf, 14, sizeof(kBuf)), kBuf + 19);
+  EXPECT_EQ(memchr_func(kBuf, 15, sizeof(kBuf)), kBuf + 20);
+  EXPECT_EQ(memchr_func(kBuf, 16, sizeof(kBuf)),
+            reverse ? kBuf + 23 : kBuf + 21);
+  EXPECT_EQ(memchr_func(kBuf, 17, sizeof(kBuf)),
+            reverse ? kBuf + 27 : kBuf + 24);
+}
+
+// Definitions of memory function reference implementations.
+
+void *ref_memcpy(void *dest, const void *src, size_t len) {
+  uint8_t *dest8 = (uint8_t *)dest;
+  uint8_t *src8 = (uint8_t *)src;
+  for (size_t i = 0; i < len; ++i) {
+    dest8[i] = src8[i];
+  }
+  return dest;
+}
+
+void *ref_memset(void *dest, int value, size_t len) {
+  uint8_t *dest8 = (uint8_t *)dest;
+  uint8_t value8 = (uint8_t)value;
+  for (size_t i = 0; i < len; ++i) {
+    dest8[i] = value8;
+  }
+  return dest;
+}
+
+enum {
+  kMemCmpEq = 0,
+  kMemCmpLt = -42,
+  kMemCmpGt = 42,
+};
+
+int ref_memcmp(const void *lhs, const void *rhs, size_t len) {
```

Comment:
can't we use the `__builtin`s for these reference implementations? If you want to keep these, do you want to add those as well?

Reply:
There are a handful reasons I opted to do this:

* When we start optimizing, I wanted to retain the dead-simple reference implementations to point to. This may be moot now that we have these parameterized unit tests.
* There isn't a corresponding builtin for each of our functions, e.g. there is no `__builtin_memrcmp`.
* I don't fully understand the semantics of the `__builtin` functions. For instance, what does  `&__builtin_memcpy` mean? I can pass that expression into the `INSTANTIATE_TEST_SUITE_P `, but inexplicably, plain `__builtin_memcpy` also works.

Reply:
I thought it would be interesting to compare our implementations (both naive and optimized) to the builtins. I think it's OK to fake the ones that don't exist. 

Would you mind extending the last point? `&f` and `f` are basically the same when assigning to a function pointer, so the behavior that you described does not come as a surprise.

Reply:
This comment should not blocks this PR though. PLMK if you think this is ready to be merged.

Reply:
> Would you mind extending the last point? &f and f are basically the same when assigning to a function pointer, so the behavior that you described does not come as a surprise.

Ah, I didn't realize that function lvalues implicitly converted to function pointers: https://en.cppreference.com/w/cpp/language/implicit_conversion

OK, I guess I don't a good justification for disliking the `__builtin` functions. I'm happy to delete a few reference implementations in favor of the `__builtin` functions.